fix(transformer/optional-chaining): fix context of optional chain call

hi-ogawa opened the pull request:
- closes https://github.com/oxc-project/oxc/issues/9498

`this.f?.()` is handled by https://github.com/oxc-project/oxc/pull/9505, but it looks like the same needs to be applied for member expressions in general, such as `repro.x.f?.()`.

Here is an output of esbuild
https://esbuild.github.io/try/#dAAwLjI1LjAALS10YXJnZXQ9ZXMyMDE1AGNvbnN0IHJlcHJvID0ge30KcmVwcm8uZj8uKCk7CnJlcHJvLnguZj8uKCk7CnJlcHJvLnhbImYiXT8uKCk7

---

This PR fixes following case:

- Not only member expressions but also arbitrary expressions can be a call context, such as:

```js
(0, repro).f?.();  
```

- same logic is required for `transform_and_join_expression`, which is used in cases, such as

```js
repro.x?.y.f?.()
```

hi-ogawa commented:
* **#9526** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9526?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9526?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9526).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

hi-ogawa commented on tasks/transform_conformance/tests/babel-plugin-transform-optional-chaining/test/fixtures/oxc/keep-this/output.js:
```diff
@@ -1,8 +1,21 @@
+(_repro$x$f = (_repro$x = repro.x).f) === null || _repro$x$f === void 0 ? void 0 : _repro$x$f.call(_repro$x);
+(_repro$x$y$f = (_repro$x$y = repro.x.y).f) === null || _repro$x$y$f === void 0 ? void 0 : _repro$x$y$f.call(_repro$x$y);
+(_repro$x2 = repro.x) === null || _repro$x2 === void 0 || (_repro$x2$f = _repro$x2.f) === null || _repro$x2$f === void 0 ? void 0 : _repro$x2$f.call(_repro$x2);
+(_repro$x3 = repro.x) === null || _repro$x3 === void 0 || (_repro$x3$y$f = _repro$x3.y.f) === null || _repro$x3$y$f === void 0 ? void 0 : _repro$x3$y$f.call(_repro$x3);
```
Hmm, this one is still incorrect :thinking: It's picking up `repro.x` instead of `repro.x.y` as context.

- input

```js
repro.x?.y.f?.();
```

- esbuild output: https://esbuild.github.io/try/#dAAwLjI1LjAALS10YXJnZXQ9ZXMyMDE1AGNvbnN0IHJlcHJvID0ge30KcmVwcm8ueD8ueS5mPy4oKTs

```js
var _a, _b, _c;
const repro = {};
(_c = (_a = repro.x) == null ? void 0 : (_b = _a.y).f) == null ? void 0 : _c.call(_b);
```

hi-ogawa replied:
This case probably needs to be handled through `transform_and_join_expression`. I fixed it by copy pasting the same logic from `transform_optional_expression`. https://github.com/oxc-project/oxc/pull/9526/commits/fb55dc4350b7de3ecdd637d34f79c37a04ae16ad I'll try to clean it up.

hi-ogawa commented on tasks/transform_conformance/tests/babel-plugin-transform-optional-chaining/test/fixtures/oxc/keep-this/input.ts:
```diff
@@ -1,5 +1,24 @@
+    this.x.y?.f?.();
+    this.x?.y?.f?.();
+    this["x"].f?.();
+    (0, this).f?.();
```
Also found another edge cases. Currently `call_context` targets a limited set of expression, but it's probably necessary to always save it to temporary binding for all chain call.

- input

```
class Repro {
  test() {
   this.f?.();
   (0, this).f?.();
  }
}

repro.f?.();
(0, repro).f?.();
```

- oxc output (same on main and this PR)

```js
var _repro$f, _repro, _f2;
class Repro {
	test() {
		var _this$f, _f;
		(_this$f = this.f) === null || _this$f === void 0 ? void 0 : _this$f.call(this);
		(_f = (0, this).f) === null || _f === void 0 ? void 0 : _f();
	}
}
(_repro$f = (_repro = repro).f) === null || _repro$f === void 0 ? void 0 : _repro$f.call(_repro);
(_f2 = (0, repro).f) === null || _f2 === void 0 ? void 0 : _f2();
```

- esbuild output https://esbuild.github.io/try/#dAAwLjI1LjAALS10YXJnZXQ9ZXMyMDE1AGNsYXNzIFJlcHJvIHsKICB0ZXN0KCkgewogICB0aGlzLmY/LigpOwogICAoMCwgdGhpcykuZj8uKCk7CiAgfQp9CgpyZXByby5mPy4oKTsKKDAsIHJlcHJvKS5mPy4oKTsK

```js
var _a, _b, _c;
class Repro {
  test() {
    var _a2, _b2, _c2;
    (_a2 = this.f) == null ? void 0 : _a2.call(this);
    (_c2 = (_b2 = (0, this)).f) == null ? void 0 : _c2.call(_b2);
  }
}
(_a = repro.f) == null ? void 0 : _a.call(repro);
(_c = (_b = (0, repro)).f) == null ? void 0 : _c.call(_b);
```

Dunqing commented:
Have you compared the output to Babel? The transformer implementation almost port from Babel.

hi-ogawa commented:
> Have you compared the output to Babel? The transformer implementation almost port from Babel.

I tested on babel repl ([link](https://babeljs.io/repl#?browsers=defaults&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=MYGwhgzhAEBKCmAHATge2gbwFDWgF3gjwAoBKTHXfACwEsIA6AMwH4GyBuS3POxgD2ZtO3GvQaCAnkPakuVMQLatZ8qr3H820lSIUaBDScuFzRBidpOrzfANoAifg4C6MvVWIAGADSLS7ma4AL5YoVjAqAB2RNDISGjQALyYwVzxKKiB6QlZgrpmGWgSRtlYRXnWIhWWpQU5mSXGZTVaRlWFuY7ObvVY3n4VAQVAA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Creact%2Cstage-2%2Cstage-3&prettier=false&targets=&version=7.26.9&externalPlugins=&assumptions=%7B%7D)) and the output of `keep-this/input.ts` matches except `? void 0 :` vs `||` difference.

I updated some conformance snapshots with `--filter` and `--override`, but I'm not sure if I've done it correctly. Please let me know if I need to fix up something :pray:

Dunqing reviewed (approved):
Great! The optional chaining plugin is very complicated and not easy to debug. Thank you for finding and fixing this bug!

Dunqing commented:
### Merge activity

* **Mar 4, 7:15 AM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

hi-ogawa committed:
fix(transformer/optional-chaining): fix context of optional chain call (#9526)

- closes https://github.com/oxc-project/oxc/issues/9498

`this.f?.()` is handled by https://github.com/oxc-project/oxc/pull/9505, but it looks like the same needs to be applied for member expressions in general, such as `repro.x.f?.()`.

Here is an output of esbuild
https://esbuild.github.io/try/#dAAwLjI1LjAALS10YXJnZXQ9ZXMyMDE1AGNvbnN0IHJlcHJvID0ge30KcmVwcm8uZj8uKCk7CnJlcHJvLnguZj8uKCk7CnJlcHJvLnhbImYiXT8uKCk7

---

This PR fixes following case:

- Not only member expressions but also arbitrary expressions can be a call context, such as:

```js
(0, repro).f?.();
```

- same logic is required for `transform_and_join_expression`, which is used in cases, such as

```js
repro.x?.y.f?.()
```
